### DISCUSSION TITLE
Add a editable SkinEditor Element List

### DISCUSSION THREAD

**Discussion by C0D3-M4513R** [ppy/osu#21098] [Ideas] [answered]
I think there is a better way, of just simply having a depth index or moving elements up or down in a list internally, to control which element is displayed above which.
I see a lot of possibilities for the skin editor.
Personally I would like to see a list of skin elements currently in use.
You could then move elements up and down the list.

Also a idea to go along with the Skin Element List is adding a Container
A Container could be manually sized or Dynamically sized (user setting), and all anchor points of contained elements would be relative to the container. Also Containers should be able to contain Containers. This could allow some nifty alignment of elements. (especially with non-statically sized elements such as #21065)

As for the design Idea, I had something in Mind similar to the `TreeStyleTabs` addon for Firefox. (screenshot below), but with osu skinning ofc. In this addon you can move entire collections as a single item, but if you pick a single item you can freely move it out of the current containers and inside of other containers.

Actually creating collections could be as simple as adding a new Button to the Right click menu, or just adding a button to the existing flip/rotate buttons, that exist on hovering on a element.
Whilst creating a collection with only a single item doesn't make much sense initially, it does make sense once you realise, that you can move elements and collections into other collections freely.

![grafik](https://user-images.githubusercontent.com/28912031/199690482-3c127143-6a7f-4842-8c8f-1efb54815011.png)

touches on #20114 and a part of #19767

**Comment by peppy** [comment, marked as answer]
This is definitely something we have planned. It could also potentially be used for things like the storyboard editor, depending on implementation.

Are you thinking of working on this? If so, I'd be happy for you to do up a proof of concept and take it from there.

**Reply by C0D3-M4513R** [reply]
I will look into this. I will definetly need some time to get used to programming ui. I am new to UI in osu and c# so it will take a little more time than my current pr.

I'm happy to try though!
I will see how difficult it is.
Hopefully I can get a initial pr ready in 2-3 weeks (will probably not even be functional, but at least contain some of the ui elements, I'm thinking of)
pr will be at #21119
